### PR TITLE
Microchip PIC

### PR DESCRIPTION
**Pull request by chostelet**
Disassembler completion and tests.
PIC Memory Mapper implementation and tests.

### CODE REVIEW

**Comment by uxmal** [comment]
Hello Christian, welcome back. Time for a big review it seems! Firstly, let's try making those appveyor and travis-CI errors go away.

**Comment by chostelet** [comment]
I've added a sample PIC database for tests. I wrongly assumed one wanting to decompile PIC binaries would have Microchip IDE software installed (which contains the original PIC database).

**Comment by chostelet** [comment]
John,

I'm having issues to solve those AppVeyor and Travis-CI failed checks.
In my own environment, the whole tests suite runs to completion.
The errors come from the fact the UnitTests are not able to find the PIC database (file 'picdb.zip') which is supposed to be copied in the 'UnitTests/bin/Debug' folder during solution's build. But I've have no way to verify this really occurs in those two test frameworks.
On my PC this copy does occur.
I'm stuck.

**Comment by uxmal** [comment]
I will be back in the house this evening to take a look and answer any
questions.  Is there a particulartime  tgat works for you so we could make
use of gitter.im?

--
John Källén - +46739914646
Sent from my mobile device
Apologies for brevity and misspellings

**Comment by chostelet** [comment]
I should be available by 21:00 (Paris time, same as Stockholm).

**Comment by chostelet** [comment]
Preferably by tomorrow, would that be OK with you ?

**Comment by uxmal** [comment]
I will be online both tonight and tomorrow, just drop a note on gitter -- j'y serai.

**Comment by chostelet** [comment]
OK.

 

--

Christian HOSTELET - Grenoble

 

De : John Källén [mailto:notifications@github.com] 
Envoyé : dimanche 7 janvier 2018 15:09
À : uxmal/reko <reko@noreply.github.com>
Cc : chostelet <christian.hostelet@free.fr>; Author <author@noreply.github.com>
Objet : Re: [uxmal/reko] Microchip PIC (#538)

 

I will be online both tonight and tomorrow, just drop a note on gitter -- j'y serai.

**Comment by chostelet** [comment]
Found way to accommodate AppVeyor, Travis CI and VS environment.

Now have issue with ````VisitArrayAccess```` method in ````IntraBlockDeadRegisters````  which is not implemented. Consequently can't execute the dataflow analysis phase for the PIC18 (using ````ArrayAccess```` in ````PIC18Rewriter````).
I've no idea how this method should be implemented.

**Comment by ptomin** [comment]
@chostelet  Try to use `MemoryAccess` instead of `ArrayAccess`
Example for pushing byte to stack:
```
Mem[sp] = byte
sp = sp + 1
```

**Comment by chostelet** [comment]
Following a chat with John, using the ArrayAccess was his suggestion. Reasons are that the PIC architecture has either no explicit stack or a limited functionality hardware stack. Essentially it is not used for parameters passing. Only for return addresses (CALL, RCALL, interrupts).

**Comment by ptomin** [comment]
I have implemented `VisitArrayAccess` of `IntraBlockDeadRegisters`  in 740f777

**Comment by chostelet** [comment]
Nice,
Shall use this implementation in my branch.

**Comment by uxmal** [comment]
@ptomin : thanks for helping @chostelet here. For some reason github is not notifying me of events in this PR.

File: src/Core/Code/InstructionTransformer.cs [hunk 1]
```diff
@@ -249,15 +249,15 @@ public virtual Expression VisitProcedureConstant(ProcedureConstant pc)
...
+        public virtual Expression VisitSegmentedAccess(SegmentedAccess access)
+        {
+            access.BasePointer = access.BasePointer.Accept(this);
+            access.EffectiveAddress = access.EffectiveAddress.Accept(this);
+            access.MemoryId = (MemoryIdentifier)access.MemoryId.Accept(this);
+            return access;
+        }
 
-		public virtual Expression VisitScopeResolution(ScopeResolution scope)
+        public virtual Expression VisitScopeResolution(ScopeResolution scope)
```
Comment (ptomin) [inline src/Core/Code/InstructionTransformer.cs]: @chostelet please do such minor style changes in separate commit(s) in branch(es) starting from `master` and then send PR(s). It makes easier to review this PR(changes will become smaller)

File: src/Core/IProcessorArchitecture.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public interface IProcessorArchitecture
         /// Creates an IEnumerable of disassembled MachineInstructions which consumes 
         /// its input from the provided <paramref name="imageReader"/>.
         /// </summary>
-        /// <remarks>This was previously an IEnumerator, but making it IEnumerable lets us use Linq expressions
+        /// <remarks>This was previously an IEnumerator, but making it IEnumerable lets us use LINQ expressions
```
Comment (ptomin) [inline src/Core/IProcessorArchitecture.cs]: @chostelet please do such minor doc changes in separate commit(s) in branch(es) starting from `master`

File: src/Core/Expressions/MemoryAccess.cs [hunk 3]
```diff
@@ -125,19 +125,54 @@ public override T Accept<T>(ExpressionVisitor<T> visit)
...
         }
-	}
+    }
+
+#if false
+
+    /// <summary>
+    /// Models an access to data memory bank for 8-bit PIC.
+    /// </summary>
+    public class BankMemoryAccess : MemoryAccess
```
Comment (ptomin) [inline src/Core/Expressions/MemoryAccess.cs]: @chostelet adding new expression class is the last resort. Consider using existing ones (e.g. `SegmentedAccess`). Have you discussed it with @uxmal?

**Comment by uxmal** [comment]
I pushed the branch `processor-architecture-customization`. Look for the word "goofy" for an example of how to use a windows form :)

**Comment by chostelet** [comment]
Windows form created. Form returns a class (PickerResult) containing the PIC name as a string and a boolean (allow decodong of Extended Execution mode for PIC18). Next step would be to know how to use this result to validate/influence the loader.

**Comment by uxmal** [comment]
Reach out to me on gitter or skype to resolve the issue of providing user options at load time. I just pushed  an update to `processor-architecture-customization` that configures the `IProcessorArchitecture` instance by calling its `LoadUserOptions` method when loading a raw file.

File: src/Arch/Microchip/Common/PICRegisterUniqueAddress.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+using System.Diagnostics;
+
+namespace Reko.Arch.MicrochipPIC.Common
+{
+    [DebuggerDisplay("{_debugDisplay,nq}")]
+    public class PICRegisterUniqueAddress :
+        IComparable<PICRegisterUniqueAddress>, IComparer<PICRegisterUniqueAddress>,
+        IEquatable<PICRegisterUniqueAddress>, IEqualityComparer<PICRegisterUniqueAddress>,
+        IComparable
+    {
```
Comment (uxmal) [inline src/Arch/Microchip/Common/PICRegisterUniqueAddress.cs]: Need a comment here to explain the purpose of this class.

File: src/Arch/Microchip/Common/PICAddresses.cs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+using Reko.Core;
+using Reko.Core.Types;
+using Reko.Core.Expressions;
+using System;
+
+namespace Reko.Arch.MicrochipPIC.Common
+{
+
+    public class PICAddress : Address
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICAddresses.cs]: Done.

File: src/Arch/Microchip/Common/PICAddresses.cs [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    /// <summary>
+    /// A PIC 21-bit program address.
+    /// </summary>
+    public class PICProgAddress : PICAddress
+    {
+
+        public const uint MAXPROGBYTADDR = 0x1FFFFFU;
+
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICAddresses.cs]: Done.

File: src/Arch/Microchip/Common/PICArchitecture.cs [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Reko.Arch.MicrochipPIC.Common
+{
+
+    public class PICArchitecture : ProcessorArchitecture
+    {
+        protected List<FlagGroupStorage> flagGroups;
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICArchitecture.cs]: Done. Should have think of it upfront :-(

File: src/Arch/Microchip/Common/PICArchitecture.cs [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+        public override Address MakeAddressFromConstant(Constant c)
+            => Address.Ptr32(c.ToUInt32());
+
+        public override Address MakeSegmentedAddress(Constant seg, Constant offset)
+            => ProcessorModel.CreateBankedAddress(seg.ToByte(), offset.ToByte());
+
+        public override Expression CreateStackAccess(IStorageBinder frame, int offset, DataType dataType)
+        {
+            throw new NotImplementedException("Microchip PIC has no explicit argument stack");
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICArchitecture.cs]: OK. Done. However still can change in the future depending on the C18 compiler used (if compiled sources).

File: src/Arch/Microchip/Common/PICArchitecture.cs [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+        /// <summary>
+        /// Gets PIC descriptor as retrieved from the Microchip Crownking database.
+        /// </summary>
+        public PIC PICDescriptor => ProcessorModel?.PICDescriptor;
+
+        /// <summary>
+        /// Gets the PIC processor state.
+        /// </summary>
+        public PICProcessorState State { get; private set; }
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICArchitecture.cs]: To be discussed. I need processor state for proper disassembly.

File: src/Arch/Microchip/Common/PICPostprocessProgram.cs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+using Reko.Core;
+using Reko.Libraries.Microchip;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Reko.Arch.MicrochipPIC.Common
+{
+    public class PICPostprocessProgram
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICPostprocessProgram.cs]: Done.

File: src/Arch/Microchip/Common/PICRegisterBitFieldStorage.cs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+        public PICRegisterBitFieldSortKey(byte bitPos, byte bitWidth)
+        {
+            BitPos = bitPos;
+            BitWidth = bitWidth;
+        }
+
+        public override int GetHashCode()
+        {
+            var hashCode = -2081872618;
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICRegisterBitFieldStorage.cs]: Corrected. Now using FNV hash code style.

File: src/Arch/Microchip/Common/PICRegisterStorage.cs [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+namespace Reko.Arch.MicrochipPIC.Common
+{
+
+    /// <summary>
+    /// Defines a PIC register (with/without named bit-fields or attached-registers).
+    /// </summary>
+    public class PICRegisterStorage : RegisterStorage,
+        IComparable<PICRegisterStorage>, IComparer<PICRegisterStorage>,
+        IEquatable<PICRegisterStorage>, IEqualityComparer<PICRegisterStorage>,
+        IComparable
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICRegisterStorage.cs]: To be discussed.  To summarize, I need a storage that can represent a whole PIC register with its bits decomposition (this is not only for STATUS register). Any PIC register can be defined with a decomposition in bits with specific individual access properties.

File: src/Arch/Microchip/Common/PICRegisterSizedUniqueAddress.cs [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Reko.Arch.MicrochipPIC.Common
+{
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Arch/Microchip/Common/PICRegisterSizedUniqueAddress.cs]: Explanation added. To be discussed if clear enough.

File: src/Arch/Microchip/Common/PICRegisterUniqueAddress.cs (hunk 4 above)
Comment (chostelet) [reply src/Arch/Microchip/Common/PICRegisterUniqueAddress.cs]: Explanation added. To be discussed if clear enough.

**Comment by uxmal** [comment]
Thanks for your efforts, this is a great contribution and a lot of work. Much appreciated!

File: src/Core/Types/PrimitiveType.cs [hunk 14]
```diff
@@ -158,6 +158,28 @@ public static PrimitiveType CreateWord(int bitSize)
 			return Create(dom, (short) bitSize, name);
 		}
 
+        //$TODO: Remove when PrimitiveType.CreateWord will accept a bit count.
+        public static PrimitiveType CreateWordFromBits(uint bitSize)
```
Comment (ptomin) [inline src/Core/Types/PrimitiveType.cs]: Shouldn't be removed?

File: src/Reko-decompiler.sln [hunk 15]
```diff
@@ -1630,8 +1650,8 @@ Global
 		{9E4B5291-EBA1-4127-AC38-E0F822F7E4F1}.TravisRelease|x64.Build.0 = TravisRelease|Any CPU
 		{9E4B5291-EBA1-4127-AC38-E0F822F7E4F1}.TravisRelease|x86.ActiveCfg = TravisRelease|Any CPU
 		{9E4B5291-EBA1-4127-AC38-E0F822F7E4F1}.TravisRelease|x86.Build.0 = TravisRelease|Any CPU
-		{1DBA55A8-6C67-4D16-8137-912B79715973}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{1DBA55A8-6C67-4D16-8137-912B79715973}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{1DBA55A8-6C67-4D16-8137-912B79715973}.Debug|Any CPU.ActiveCfg = Debug|x86
+		{1DBA55A8-6C67-4D16-8137-912B79715973}.Debug|Any CPU.Build.0 = Debug|x86
```
Comment (ptomin) [inline src/Reko-decompiler.sln]: Is it correct?

File: src/packages/NUnit.2.6.4/lib/nunit.framework.xml [hunk 16]
```diff
@@ -0,0 +1,10984 @@
+<?xml version="1.0"?>
+<doc>
```
Comment (ptomin) [inline src/packages/NUnit.2.6.4/lib/nunit.framework.xml]: Was `NUnit.2.6.4` folder added by mistake?

File: src/Gui/Forms/OpenAsInteractor.cs [hunk 17]
```diff
@@ -60,6 +61,7 @@ private void dlg_Load(object sender, EventArgs e)
         private void EnableControls()
         {
             var rawfile = ((ListOption)dlg.RawFileTypes.SelectedValue).Value as RawFileElement;
+            var arch = ((ListOption)dlg.Architectures.SelectedValue)?.Value as string;
```
[the remark this answers is not in the document]
Comment (chostelet) [reply src/Gui/Forms/OpenAsInteractor.cs]: I can't comment on this part.

[the remark this answers is not in the document]
Comment (chostelet) [reply src/Gui/Forms/OpenAsInteractor.cs]: No. Maybe some typos or tabs were modified (not on purpose) but I'm not the genitor of this OpenAsInteractor.cs. But you're right, 'arch' is unused in this method.

File: src/Core/Types/PrimitiveType.cs (hunk 14 above)
Comment (uxmal) [reply src/Core/Types/PrimitiveType.cs]: Let's have both a `uint` and an `int` version of `PrimitiveType.CreateWord` for convenience. C# is stricter than C / C++ when it comes to unsignedness and all those casts are a pain.

File: src/Gui/Forms/OpenAsInteractor.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Gui/Forms/OpenAsInteractor.cs]: That's my code. The code is there in preparation for a more complete implementation once we provide options capability for all fields in the `Open As...` dialog in addition to architectures.

File: src/packages/NUnit.2.6.4/lib/nunit.framework.xml (hunk 16 above)
Comment (chostelet) [reply src/packages/NUnit.2.6.4/lib/nunit.framework.xml]: That was probably not the intention. The only (kind-of) explanation I may have, is that when I tried to clone Reko repository on a new machine, the nunit.framwork.dll was missing and was unable to execute a vast majority of the tests. Probably the result of a NuGet action.

File: src/Core/Types/PrimitiveType.cs (hunk 14 above)
Comment (chostelet) [reply src/Core/Types/PrimitiveType.cs]: OK, shall do in next PR.

File: src/Reko-decompiler.sln (hunk 15 above)
Comment (chostelet) [reply src/Reko-decompiler.sln]: Shall review all these AnyCPU and specific x86 configuration settings

File: src/Core/Types/PrimitiveType.cs (hunk 14 above)
Comment (chostelet) [reply src/Core/Types/PrimitiveType.cs]: What shall we do, if by chance, int bitSize is <= 0? Negate? Make it bool (forcing to 1)?

File: src/Reko-decompiler.sln (hunk 15 above)
Comment (uxmal) [reply src/Reko-decompiler.sln]: I'll take a peek at the SLN file.

File: src/Core/Types/PrimitiveType.cs (hunk 14 above)
Comment (uxmal) [reply src/Core/Types/PrimitiveType.cs]: Throw `ArgumentException` for negative sizes

Comment (uxmal) [reply src/Core/Types/PrimitiveType.cs]: Or as you say ArgumentOutOfRangeException

File: src/packages/NUnit.2.6.4/lib/nunit.framework.xml (hunk 16 above)
Comment (uxmal) [reply src/packages/NUnit.2.6.4/lib/nunit.framework.xml]: We should consider adding Nunit as a nuget package though. I can tinker with this this evening

Comment (ptomin) [reply src/packages/NUnit.2.6.4/lib/nunit.framework.xml]: > That was probably not the intention.

So it should be removed. It works fine without that directory on my machine (and on the `Travis`/'Appveyor')

File: src/Gui/Forms/OpenAsInteractor.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Gui/Forms/OpenAsInteractor.cs]: @chostelet  you have added this line in a286d7c. So it's your code

File: subjects/Microchip/C18/PIC18Extd/subject.cmd [hunk 18]
```diff
@@ -0,0 +1 @@
+decompile.exe --heuristic shingle --arch pic --arch_option model="PIC18F1230" --loader IntelHex --base 0000 --entry 0000 PIC18EggExtd.hex
```
Comment (ptomin) [inline subjects/Microchip/C18/PIC18Extd/subject.cmd]: Should be `--arch-option` instead of `--arch_option`

File: subjects/Microchip/C18/PIC18/subject.cmd [hunk 19]
```diff
@@ -0,0 +1 @@
+decompile.exe --heuristic shingle --arch pic --arch_option model="PIC18F1220" --loader IntelHex --base 0000 --entry 0000 PIC18Legacy.hex
```
Comment (ptomin) [inline subjects/Microchip/C18/PIC18/subject.cmd]: Should be `--arch-option` too

File: src/Reko-decompiler.sln (hunk 15 above)
Comment (uxmal) [reply src/Reko-decompiler.sln]: This is the NativeProxy assembly, whose `Any CPU` target is meaningless. NativeProxy exists solely to pass on whether or not x86 or x64 is the target platform to the CMake build which compiles the native components.  The real fix is to remove the _solution_ platform `Any CPU`, since we can't build Reko "platform neutrally" because of the dependency on native binaries.

I will investigate what happens when we remove "Any CPU" from the set of available solution platforms. Note that individual _project_ platforms for the majority of the C# projects will remain `Any CPU`.

**Comment by uxmal** [comment]
@ptomin: my fix for #642 should resolve the issue where `PIC18EggExtd.dcproject` is failing to load.

File: src/Core/Types/PrimitiveType.cs (hunk 14 above)
Comment (uxmal) [reply src/Core/Types/PrimitiveType.cs]: In fact, throw `ArgumentOutOfRangeException` for `bitSize <= 0` in `CreateWord(int bitSize)`.

**Comment by ptomin** [comment]
But it does not. See https://ci.appveyor.com/project/uxmal/reko/build/0.8.0.988

File: subjects/Microchip/C18/PIC18Extd/subject.cmd (hunk 18 above)
Comment (chostelet) [reply subjects/Microchip/C18/PIC18Extd/subject.cmd]: Shall change this misspelled argument. 
No default model yet. Can think of it - likewise what is defined by Microchip for their COFF format (namely PIC18F4620).

File: src/Core/Types/PrimitiveType.cs (hunk 14 above)
Comment (chostelet) [reply src/Core/Types/PrimitiveType.cs]: This is what I'll implement. However, I'm getting regression using those CreateWord() due to the fact there is no corresponding Constant. (ex. I'm having 24-bit type for some PIC registers but I can't defined 24-bit Constant by using current Constant.Create.
